lib/src/set_multimap/built_set_multimap.dart

```diff
@@ -0,0 +1,191 @@
+      (toBuilder()..update(updates)).build();
+
+  /// Converts to a [Map].
+  ///
+  /// Note that the implementation is efficient: it returns a copy-on-write
+  /// wrapper around the data from this `BuiltSetMultimap`. So, if no mutations are
+  /// made to the result, no copy is made.
+  ///
+  /// This allows efficient use of APIs that ask for a mutable collection
+  /// but don't actually mutate it.
+  Map<K, BuiltSet<V>> toMap() => new CopyOnWriteMap<K, BuiltSet<V>>(_map);
+
+  /// Deep hashCode.
+  ///
+  /// A `BuiltSetMultimap` is only equal to another `BuiltSetMultimap` with
+  /// equal key/values pairs in any order. Then, the `hashCode` is guaranteed
+  /// to be the same.
+  @override
+  int get hashCode {
+    if (_hashCode == null) {
+      _hashCode = hashObjects(_map.keys
+          .map((key) => hash2(key.hashCode, _map[key].hashCode))
+          .toList(growable: false)..sort());
+    }
+    return _hashCode;
+  }
+
+  /// Deep equality.
+  ///
+  /// A `BuiltSetMultimap` is only equal to another `BuiltSetMultimap` with
+  /// equal key/values pairs in any order.
+  @override
+  bool operator ==(other) {
+    if (other is! BuiltSetMultimap) return false;
+    if (other.length != length) return false;
+    if (other.hashCode != hashCode) return false;
+    for (final key in keys) {
+      if (other[key] != this[key]) return false;
+    }
+    return true;
+  }
+
+  String toString() => _map.toString();
+
+  // SetMultimap.
+
+  /// As [SetMultimap], but results are [BuiltSet]s and not mutable.
+  BuiltSet<V> operator [](K key) {
+    final result = _map[key];
+    return identical(result, null) ? _emptySet : result;
+  }
+
+  /// As [SetMultimap.containsKey].
+  bool containsKey(Object key) => _map.containsKey(key);
+
+  /// As [SetMultimap.containsValue].
+  bool containsValue(Object value) => values.contains(value);
+
+  /// As [SetMultimap.forEach].
+  void forEach(void f(K key, V value)) {
+    _map.forEach((key, values) {
+      values.forEach((value) {
+        f(key, value);
+      });
+    });
+  }
+
+  /// As [SetMultimap.forEachKey].
+  void forEachKey(void f(K key, Iterable<V> value)) {
+    _map.forEach((key, values) {
+      f(key, values);
+    });
+  }
+
+  /// As [SetMultimap.isEmpty].
+  bool get isEmpty => _map.isEmpty;
+
+  /// As [SetMultimap.isNotEmpty].
+  bool get isNotEmpty => _map.isNotEmpty;
+
+  /// As [SetMultimap.keys], but result is stable; it always returns the same
+  /// instance.
+  Iterable<K> get keys {
+    if (_keys == null) {
+      _keys = _map.keys;
+    }
+    return _keys;
+  }
+
+  /// As [SetMultimap.length].
+  int get length => _map.length;
+
+  /// As [SetMultimap.values], but result is stable; it always returns the
+  /// same instance.
+  Iterable<V> get values {
+    if (_values == null) {
+      _values = _map.values.expand((x) => x);
+    }
+    return _values;
+  }
+
+  // Internal.
+
+  BuiltSetMultimap._copyAndCheck(Iterable keys, Function lookup)
```

Comment:
This is an implementation detail, I didn't want it cluttering up the public / interesting methods at the top.